crates/xilem_masonry/src/any_view.rs

```diff
@@ -0,0 +1,177 @@
+use std::ops::Deref;
+
+use masonry::{
+    declare_widget,
+    widget::{StoreInWidgetMut, WidgetMut, WidgetRef},
+    BoxConstraints, EventCtx, LayoutCtx, LifeCycle, LifeCycleCtx, PaintCtx, Point, PointerEvent,
+    Size, StatusChange, TextEvent, Widget, WidgetPod,
+};
+use smallvec::SmallVec;
+use vello::Scene;
+
+use crate::{ChangeFlags, MasonryView, MessageResult, ViewCx, ViewId};
+
+/// A view which can have any underlying view type.
+///
+/// This can be used to return type erased views (such as from a trait),
+/// or used to implement conditional display and switching of views.
+///
+/// Note that `Option` can also be used for conditionally displaying
+/// views when placed in a ViewSequence.
+// TODO: Mention `Either` when we have implemented that?
+pub type BoxedMasonryView<T, A = ()> = Box<dyn AnyMasonryView<T, A>>;
+
+impl<T: 'static, A: 'static> MasonryView<T, A> for BoxedMasonryView<T, A> {
+    type Element = DynWidget;
+
+    fn build(&self, cx: &mut ViewCx) -> masonry::WidgetPod<Self::Element> {
+        self.deref().dyn_build(cx)
+    }
+
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut T,
+    ) -> crate::MessageResult<A> {
+        self.deref().dyn_message(id_path, message, app_state)
+    }
+
+    fn rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &Self,
+        // _id: &mut Id,
+        element: masonry::widget::WidgetMut<Self::Element>,
+    ) -> ChangeFlags {
+        self.deref().dyn_rebuild(cx, prev, element)
+    }
+}
+
+/// A trait enabling type erasure of views.
+pub trait AnyMasonryView<T, A = ()>: Send {
+    fn as_any(&self) -> &dyn std::any::Any;
+
+    fn dyn_build(&self, cx: &mut ViewCx) -> WidgetPod<DynWidget>;
+
+    fn dyn_rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &dyn AnyMasonryView<T, A>,
+        element: WidgetMut<DynWidget>,
+    ) -> ChangeFlags;
+
+    fn dyn_message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn std::any::Any>,
+        app_state: &mut T,
+    ) -> MessageResult<A>;
+}
+
+impl<T, A, V: MasonryView<T, A> + 'static> AnyMasonryView<T, A> for V {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn dyn_build(&self, cx: &mut ViewCx) -> WidgetPod<DynWidget> {
+        let element = self.build(cx);
+        WidgetPod::new(DynWidget {
+            inner: element.boxed(),
+        })
+    }
+
+    fn dyn_rebuild(
+        &self,
+        cx: &mut ViewCx,
+        prev: &dyn AnyMasonryView<T, A>,
+        mut element: WidgetMut<DynWidget>,
+    ) -> ChangeFlags {
+        // TODO: Does this need to have a custom view id to enable events sent
+        // to an outdated view path to be caught and returned?
+        // Should we store this generation in `element`? Seems plausible
```

Comment:
It's not really decoupled from the widgets then, is it? (Which AFAIK is necessary, for views not being wired up directly with widgets but being able to receive events from somewhere else, async list is an example for that (via the waker that stores the current view id path)).
I think it's one of the reasons why `View::build` returns a `ViewId`, and rebuild is able to change it (e.g. in this method, because there's a new view).

Reply:
Yeah, this does need to happen in `View::State`.

This isn't actually problematic because we don't have async wiring yet, but the current state is wrong